File: src/main/java/com/atsuishio/superbwarfare/init/ModSerializers.java
package com.atsuishio.superbwarfare.init;

import com.atsuishio.superbwarfare.Mod;
import com.atsuishio.superbwarfare.data.gun.GunData;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.syncher.EntityDataSerializer;
import net.neoforged.neoforge.registries.DeferredHolder;
import net.neoforged.neoforge.registries.DeferredRegister;
import net.neoforged.neoforge.registries.NeoForgeRegistries;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ModSerializers {

    public static final DeferredRegister<EntityDataSerializer<?>> REGISTRY = DeferredRegister.create(NeoForgeRegistries.Keys.ENTITY_DATA_SERIALIZERS, Mod.MODID);

    public static final DeferredHolder<EntityDataSerializer<?>, EntityDataSerializer<List<Integer>>> INT_LIST_SERIALIZER = REGISTRY.register("int_list_serializer",
            () -> EntityDataSerializer.forValueType(ByteBufCodecs.VAR_INT.apply(ByteBufCodecs.list()))
    );
    public static final DeferredHolder<EntityDataSerializer<?>, EntityDataSerializer<List<Float>>> FLOAT_LIST_SERIALIZER = REGISTRY.register("float_list_serializer",
            () -> EntityDataSerializer.forValueType(ByteBufCodecs.FLOAT.apply(ByteBufCodecs.list()))
    );

    public static final DeferredHolder<EntityDataSerializer<?>, EntityDataSerializer<Map<String, GunData>>> VEHICLE_GUN_DATA_MAP_SERIALIZER = REGISTRY.register("vehicle_gun_data_map_serializer",
            () -> new EntityDataSerializer<>() {
                @Override
                public @NotNull StreamCodec<? super RegistryFriendlyByteBuf, Map<String, GunData>> codec() {
                    return ByteBufCodecs.map(HashMap::new, ByteBufCodecs.STRING_UTF8, GunData.VEHICLE_GUN_STREAM_CODEC);
                }

                public @NotNull Map<String, GunData> copy(@NotNull Map<String, GunData> map) {
                    var newMap = new HashMap<String, GunData>();
                    map.forEach((key, value) -> newMap.put(key, value.copy()));
                    return newMap;
                }
            });
}

File: src/main/kotlin/com/atsuishio/superbwarfare/init/ModSerializers.kt
package com.atsuishio.superbwarfare.init

import com.atsuishio.superbwarfare.Mod
import com.atsuishio.superbwarfare.data.gun.GunData
import net.minecraft.network.syncher.EntityDataSerializer
import net.minecraft.world.item.ItemStack
import net.minecraftforge.registries.DeferredRegister
import net.minecraftforge.registries.ForgeRegistries
import net.minecraftforge.registries.RegistryObject

object ModSerializers {
    val REGISTRY: DeferredRegister<EntityDataSerializer<*>> =
        DeferredRegister.create(ForgeRegistries.Keys.ENTITY_DATA_SERIALIZERS, Mod.MODID)

    @JvmField
    val INT_LIST_SERIALIZER: RegistryObject<EntityDataSerializer<MutableList<Int>>> =
        REGISTRY.register("int_list_serializer") {
            EntityDataSerializer.simple({ buf, list ->
                buf.writeVarInt(list.size)
                for (v in list) {
                    buf.writeVarInt(v)
                }
            }, { buf ->
                val length = buf.readVarInt()
                val list = arrayListOf<Int>()
                repeat(length) {
                    list.add(buf.readVarInt())
                }
                list
            })
        }

    @JvmField
    val FLOAT_LIST_SERIALIZER: RegistryObject<EntityDataSerializer<MutableList<Float>>> =
        REGISTRY.register("float_list_serializer") {
            EntityDataSerializer.simple({ buf, list ->
                buf.writeVarInt(list.size)
                for (v in list) {
                    buf.writeFloat(v)
                }
            }, { buf ->
                val length = buf.readVarInt()
                val list = arrayListOf<Float>()
                repeat(length) {
                    list.add(buf.readFloat())
                }
                list
            })
        }

    @JvmField
    val VEHICLE_GUN_DATA_MAP_SERIALIZER: RegistryObject<EntityDataSerializer<MutableMap<String, GunData>>> =
        REGISTRY.register("vehicle_gun_data_map_serializer") {
            EntityDataSerializer.simple({ buf, map ->
                buf.writeVarInt(map.size)
                for (kv in map.entries) {
                    buf.writeUtf(kv.key)
                    buf.writeNbt(kv.value.stack.shareTag)
                }
            }, { buf ->
                val length = buf.readVarInt()
                val map = hashMapOf<String, GunData>()
                repeat(length) {
                    val weaponName = buf.readUtf()

                    val tag = buf.readNbt()
                    val gunItemStack = ItemStack(ModItems.VEHICLE_GUN.get())
                    gunItemStack.setTag(tag)

                    map[weaponName] = GunData.from(gunItemStack)
                }
                map
            })
        }
}